shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm

```diff
@@ -79,72 +82,81 @@ static bool ClipRRectContainsPlatformViewBoundingRect(const SkRRect& clip_rrect,
-                                                               pixelFormat:pixel_format]);
-      overlay_view_wrapper.reset([[FlutterOverlayView alloc] initWithContentsScale:screenScale
-                                                                       pixelFormat:pixel_format]);
-
-      auto ca_layer = fml::scoped_nsobject<CALayer>{[overlay_view.get() layer]};
-      std::unique_ptr<IOSSurface> ios_surface = IOSSurface::Create(ios_context, ca_layer);
-      std::unique_ptr<Surface> surface = ios_surface->CreateGPUSurface(gr_context);
-
-      layer = std::make_shared<FlutterPlatformViewLayer>(
-          std::move(overlay_view), std::move(overlay_view_wrapper), std::move(ios_surface),
-          std::move(surface));
-      layer->gr_context = gr_context;
-    }
-    // The overlay view wrapper masks the overlay view.
-    // This is required to keep the backing surface size unchanged between frames.
-    //
-    // Otherwise, changing the size of the overlay would require a new surface,
-    // which can be very expensive.
-    //
-    // This is the case of an animation in which the overlay size is changing in every frame.
-    //
-    // +------------------------+
-    // |   overlay_view         |
-    // |    +--------------+    |              +--------------+
-    // |    |    wrapper   |    |  == mask =>  | overlay_view |
-    // |    +--------------+    |              +--------------+
-    // +------------------------+
-    layer->overlay_view_wrapper.get().clipsToBounds = YES;
-    [layer->overlay_view_wrapper.get() addSubview:layer->overlay_view];
-    layers_.push_back(layer);
+std::shared_ptr<FlutterPlatformViewLayer> FlutterPlatformViewLayerPool::GetNextLayer() {
+  if (available_layer_index_ < layers_.size()) {
+    // // TODO: Skia
+    std::shared_ptr<FlutterPlatformViewLayer> layer = layers_[available_layer_index_];
+
+    // // This condition can only happen with the Skia backend, which is due to be removed from
+    // // iOS in short order.
+    // if (gr_context != layer->gr_context) {
+    //   layer->gr_context = gr_context;
+    //   // The overlay already exists, but the GrContext was changed so we need to recreate
+    //   // the rendering surface with the new GrContext.
+    //   IOSSurface* ios_surface = layer->ios_surface.get();
+    //   std::unique_ptr<Surface> surface = ios_surface->CreateGPUSurface(gr_context);
+    //   layer->surface = std::move(surface);
+    // }
+
+    available_layer_index_++;
+    return layer;
   }
-  std::shared_ptr<FlutterPlatformViewLayer> layer = layers_[available_layer_index_];
-  if (gr_context != layer->gr_context) {
-    layer->gr_context = gr_context;
-    // The overlay already exists, but the GrContext was changed so we need to recreate
-    // the rendering surface with the new GrContext.
-    IOSSurface* ios_surface = layer->ios_surface.get();
+
+  return nullptr;
+}
+
+void FlutterPlatformViewLayerPool::CreateLayer(GrDirectContext* gr_context,
```

Comment:
This is probably the biggest PITA. Because I can only construct these on the platform thread, right now frame 0 with platform views will skip some number of overlay layers. Its possible that the FlutterMetalLayer solves this partially

Reply:
We we move away from `CAMetalLayer` (or even `FlutterMetalLayer`) and adopt custom composition, we can solve this easily because you can allocate `IOSurfaces` on any thread, when they are not tied to particular `CALayer`. We can also have much cheaper overlays. Right now we use up to 2 `CAMetaLayer`s for each `FlutterView` overlay, which is 6 full screen surfaces. With custom composition it is possible to set a single iOSurface as content of as many `CALayer`s as needed each showing different part, so you can many unobstructed platform view overlays with single layer.

Surfaces are [allocated on raster thread](https://github.com/flutter/engine/blob/c0066709e6fb27bef1cf618e7c65a6fe3a3cde40/shell/platform/darwin/macos/framework/Source/FlutterSurfaceManager.mm#L133) and the overlay sublayers are crated [here](https://github.com/flutter/engine/blob/c0066709e6fb27bef1cf618e7c65a6fe3a3cde40/shell/platform/darwin/macos/framework/Source/FlutterSurfaceManager.mm#L63). 

Here are relevant files:

https://github.com/flutter/engine/blob/main/shell/platform/darwin/macos/framework/Source/FlutterSurfaceManager.mm

https://github.com/flutter/engine/blob/main/shell/platform/darwin/macos/framework/Source/FlutterCompositor.mm